src/managers/tls.py

```diff
@@ -52,37 +52,77 @@ def set_certificate(self) -> None:
 
     def set_truststore(self) -> None:
         """Creates the unit Java Truststore and adds the unit CA."""
-        keytool_cmd = "charmed-zookeeper.keytool" if self.substrate == "vm" else "keytool"
-
         try:
-            self.workload.exec(
-                command=[
-                    keytool_cmd,
-                    "-import",
-                    "-v",
-                    "-alias",
-                    "ca",
-                    "-file",
-                    self.workload.paths.ca,
-                    "-keystore",
-                    self.workload.paths.truststore,
-                    "-storepass",
-                    self.state.unit_server.truststore_password,
-                    "-noprompt",
-                ],
-            )
+            self._import_ca_in_truststore()
+
             if self.substrate == "vm":
                 self.workload.exec(
                     command=["chown", "snap_daemon:root", self.workload.paths.truststore],
                 )
-
         except (subprocess.CalledProcessError, ops.pebble.ExecError) as e:
             if "already exists" in str(e.stdout):
+                # Replacement strategy:
+                # - We need to own the file, otherwise keytool throws a permission error upon removing an entry
+                # - We need to make sure that the keystore is not empty at any point, hence the four steps.
+                #  Otherwise, ZK would pick up the file change when it's empty, and crash its internal watcher thread
+                try:
+                    if self.substrate == "vm":
+                        self.workload.exec(
+                            command=["chown", "root:root", self.workload.paths.truststore],
+                        )
+                    self._import_ca_in_truststore("ca-upcoming")
+                    self._delete_ca_in_truststore("ca")
+                    self._import_ca_in_truststore("ca")
+                    self._delete_ca_in_truststore("ca-upcoming")
```

Comment:
**suggestion**: My preferred way to do this would be to always have two certs and do this in just two steps.
1) Grab current alias "ca" and **replace**/rename to "old-ca"
```
keytool -changealias -alias "ca" -destalias "old-ca" . . .
```
2) Import and replace new cert with alias "ca"

As far as I know, there is no problem with keeping the expired one around.

Reply:
Ah, I think -changealias would also throw an `already exists` exception. In that case the flow would be:
1) Delete `old-ca` alias if it exists
2) `-changealias` from current `ca` to `old-ca`
3) Import new cert with alias `ca`

Reply:
That's a great suggestion, thank you. IMO it would be best to always remove the old ca (so basically, move your step 1 to last position) because (1) we do not want a client to continue using a revoked certificate and (2) "if it exists" adds another layer of exception handling. I'll give it a try first

Reply:
Yes, I confirm changing the alias throws the same exception. 3 steps replacement added in 8f6cb572a32e

Reply:
I believe the client will always use the certificate that works. If we have an expired cert on the truststore, the service will simply not use it